### PR TITLE
roachtest: replace roachtest run --dry-run

### PR DESCRIPTION
**Pull request by andreimatei**
... with a dedicated `roachtest list`.
The dryrun argument was polluting and indenting the code.

Release note: None

### CODE REVIEW

**Comment by andreimatei** [comment]
The first commit is #30146.

**Review by petermattis — CHANGES_REQUESTED** [review]
This unfortunately removes functionality: `roachtest run -n` vs `roachtest bench -n`.

> Reviewed 4 of 4 files at r1.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/30994#-:-LO0myH-9J0q141tWEpV:bpzff5j)* status: :shipit: complete! 0 of 0 LGTMs obtained

___
*[pkg/cmd/roachtest/main.go, line 43 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/30994#-LO0mYDN6n55M4Gi4_fr:-LO0mYDO7UdtzBdqiwe4:b-69z9nk) ([raw file](https://github.com/cockroachdb/cockroach/blob/de6cff5936b26d8962df47822f246e62e7d0b19e/pkg/cmd/roachtest/main.go#L43)):*
> ```Go
> 				return fmt.Errorf("cannot specify both an existing cluster (%s) and --local", clusterName)
> 			}
> 			if !dryrun {
> ```

This was useful, though, as it prevented trying to find the `roachprod` and `workload` binaries if you weren't running a test. I think you'll want to refactor where `initBinaries` is performed so that it is only done if we're actually running tests.

**Review by andreimatei — COMMENTED** [review]
`roachtest bench`... who knew. Added `roachtest list --bench`.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/30994#-:-LO15BruEKY6OC-BKqXD:b7j7ziu)* status: :shipit: complete! 0 of 0 LGTMs obtained

___
*[pkg/cmd/roachtest/main.go, line 43 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/30994#-LO0mYDN6n55M4Gi4_fr:-LO156kIB5K_J44DXpLJ:b-qlw9di) ([raw file](https://github.com/cockroachdb/cockroach/blob/de6cff5936b26d8962df47822f246e62e7d0b19e/pkg/cmd/roachtest/main.go#L43)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary><blockquote>

This was useful, though, as it prevented trying to find the `roachprod` and `workload` binaries if you weren't running a test. I think you'll want to refactor where `initBinaries` is performed so that it is only done if we're actually running tests.
</blockquote></details>

good catch. I left it here, but made it conditional on the command.

**Review by petermattis — COMMENTED** [review]
I think it is good that you got rid of the `dryrun` variable infecting the `run` implementation. But I'm less thrilled about the user ergonomics of `list` and `list --bench`. A `--dry-run` flag is reasonably common in tools. Did you consider leaving the flag in place, but changing the implementation to be similar to what you've implemented for `listCmd`? That is, the `runCmd` function would contain something like:

```
if dryrun {
	names := r.ListAll(args)
	for _, name := range names {
		fmt.Println(name)
	}
} else {
	os.Exit(r.Run(args))
}
```

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/30994#-:-LO3L7FACNSwwyjs0jty:bpgyjcb)* status: :shipit: complete! 0 of 0 LGTMs obtained

___
*[pkg/cmd/roachtest/main.go, line 48 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/30994#-LO3Kroa5piYJTWCW38d:-LO3Kroa5piYJTWCW38e:bxe13bd) ([raw file](https://github.com/cockroachdb/cockroach/blob/bb28afe95bcc7135b981acf70831198f42204b81/pkg/cmd/roachtest/main.go#L48)):*
> ```Go
> 				"store-gen": {},
> 			}
> 			if _, ok := needsBinaries[cmd.Name()]; ok {
> ```

Nit: I think this is cleaner as a `switch`:

```
switch cmd.Name() {
case "run", "bench", "store-gen":
  initBinaries()
}
```

**Review by andreimatei — COMMENTED** [review]
I have considered it.
`--dry-run` may be common enough for simple tools, but I don't think `roachtest` qualifies (particularly as I'm trying to make it more complex). This tool should allocate a cluster pool, then run tests through a scheduler. Were a `--dry-run` to exist, it'd be unclear (to me as a user) what parts should run and what parts shouldn't: how would it interact with the --parallelism and --count flags, for example?
Also, through the `run()` decontamination, I've also changed the output format of `list/--dry-run` - it no longer prints stuff like `[PASS]` which is a big deal as far as expectations from this tool go.
Basically, to quote a living classic, to consider `--dry-run` making sense with `run` I have to squint so hard my eyes would be closed.

Having said all this, I'll change it if you insist.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/30994#-:-LO3PVMM7eRHbmFDnord:b-3rt1ke)* status: :shipit: complete! 0 of 0 LGTMs obtained

**Review by petermattis — APPROVED** [review]
I don't feel strongly about `--dry-run` and you raise good points. Another approach would be to replace `--dry-run` with `--list`. That is `run --list`, `bench --list`, etc. Feel free to ignore this suggestion.

:lgtm:

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/30994#-:-LO3a-tU4lm7Fw14p2ox:b-3e2dc2)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale)

**Review by andreimatei — COMMENTED** [review]
I... still prefer a top-level command (particularly as you have a good suggestion elsewhere about also optionally printing the cluster spec for each test, which I might do).

Merging on Monday.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/30994#-:-LO4b0eMBvXVc3i9z8xG:b-7jgiu4)* status: :shipit: complete! 1 of 0 LGTMs obtained

___
*[pkg/cmd/roachtest/main.go, line 48 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/30994#-LO3Kroa5piYJTWCW38d:-LO4azLADXEZOfbTAOEC:bb7f2mf) ([raw file](https://github.com/cockroachdb/cockroach/blob/bb28afe95bcc7135b981acf70831198f42204b81/pkg/cmd/roachtest/main.go#L48)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary><blockquote>

Nit: I think this is cleaner as a `switch`:

```
switch cmd.Name() {
case "run", "bench", "store-gen":
  initBinaries()
}
```
</blockquote></details>

done

**Review by andreimatei — COMMENTED** [review]
bors r+

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/30994#-:-LOJMxmT3L_hlIqlrJ0y:b-ydi9ay)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale)
